### DISCUSSION TITLE
How to use os.version and os.features in Linux and are they really forced upon pulling images?

### DISCUSSION THREAD

**Discussion by aslanpour** [moby/moby#43871] [Q&A]
Hi,

I have to achieve a **Docker manifest** that contains 2 images for **identical (Linux) OS and Arch** but for **different OS versions and features**. I thought maybe I can annotate each image with different values for `os.version` and `os.features` so when pulling the image, the one relevant to the host's os version and features gets pulled. 
Question:

- **Where to obtain the correct value for os.version and os.features in Linux?** My guess is that they come from cat /etc/os-release and cat /proc/cpuinfo. 

- **Are the annotations strictly checked by Docker when pulling an image of a manifest or they are just informational?**

An example docker manifest in my mind is this:
Base image -->  foo/bar:v1
image1       -->   foo/bar:os1         
image2       --> foo/bar: os2          

image1 annotation will be os.version X  os.features XX       where X and XX refer to host1 information.
image2 annotation will be os.version Y os.features YY         where Y and YY refer to host2 information.

By creating and pushing this manifest, I would like Docker to pull the relevant image to a host upon running
docker run foo/bar:v1
E.g., if the host os version and features match X and XX, pull image1.

Note that I am just concerned about Linux systems.

 My host1 information for cat /etc/os-release is as follows
```
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

I really appreciate your help.
I can provide a detailed problem if required.
(I feel the best would be to enable environmental-variable-based pulling so any use case can benefit)

**Comment by thaJeztah** [comment]
The "os.features" and "os.version" options are a bit of a gray area in the OCI specification. These fields are currently only used to match Windows images, where `os.version` is important to match a compatible kernel version (and to some extend `os.features` may be used as well).

On Linux, all current implementations of OCI runtimes use `os` (i.e., `linux`) `arch` (e.g. `amd64`) and `variant` (mostly for ARM architectures, but some x86 variants _may_ be used.

From the OCI image-spec; https://github.com/opencontainers/image-spec/blob/02efb9a75ee11e05937b535cc5f228f9343ab2f5/image-index.md#L68


- **`os.version`** *string*

    This OPTIONAL property specifies the version of the operating system targeted by the referenced blob.
    Implementations MAY refuse to use manifests where `os.version` is not known to work with the host OS version.
    Valid values are implementation-defined. e.g. `10.0.14393.1066` on `windows`.

- **`os.features`** *array of strings*

    This OPTIONAL property specifies an array of strings, each specifying a mandatory OS feature.
    When `os` is `windows`, image indexes SHOULD use, and implementations SHOULD understand the following values:

    - `win32k`: image requires `win32k.sys` on the host (Note: `win32k.sys` is missing on Nano Server)

    When `os` is not `windows`, values are implementation-defined and SHOULD be submitted to this specification for standardization.
